### ISSUE TITLE
Improved matrix solver

### ISSUE THREAD

**Issue by dcbroad3** [factoriolab/factorio-lab#201]
The current cassowary matrix solver throws an error if a circular reference is found in the recipes. To prevent this from happening code prevents circular recipes from being added to the matrix solver. 
Instead, the matrix solver should handle circular recipes. A new library should be used, or written, or the current solver should be modified.

**Comment by dcbroad3** [comment]
If circular recipes are handled better by the matrix solver then they also need representation in the sankey diagram. The published d3 sankey library does not allow circular references, either use d3-sankey-circular or develop a custom implementation.

**Comment by bbarry** [comment]
I wondered if this was going to be a problem. 

It seemed so straightforward to use cassowary but I was having a hard time imagining it working for cases like BA sulfur loops where some of the intermediates may have multiple possible solutions.  In game helmod seems to get these wrong also (I've got a number of factory loops that helmod insists are 0 factories where it fails to calculate).

**Comment by dcbroad3** [comment]
@bbarry why didn't you warn me! 😄
I've spent a lot of time researching packages out there and I've been very disappointed with what I've found. I'll most likely just sit down and spend the time to develop a custom Simplex algorithm using the Rational class in Factorio Lab. That's apparently the algorithm that Kirk McDonald uses and it seems to be okay with circular recipes, it also generally seems to be the most widely used algorithm. Just hope it can be efficient.
